grails-app/services/org/pih/warehouse/data/ProductSupplierService.groovy

```diff
@@ -355,4 +357,36 @@ class ProductSupplierService {
     void delete(String productSupplierId) {
         productSupplierGormService.delete(productSupplierId)
     }
+
+    private static void validateProductSupplierDetails(ProductSupplierDetailsCommand command) {
+        if (command.hasErrors()) {
+            throw new ValidationException("Invalid product source", command.errors)
+        }
+    }
+
+    ProductSupplier saveDetails(ProductSupplierDetailsCommand command) {
+        validateProductSupplierDetails(command)
+        ProductSupplier productSupplier = new ProductSupplier(command.properties)
+        // TODO: To be replaced by a generic method created in OBPIH-6017 after merging into develop
+        if (!productSupplier.code) {
+            productSupplier.code =
+                identifierService.generateProductSupplierIdentifier(command?.product?.productCode, command?.supplier?.code)
+        }
+        return productSupplierGormService.save(productSupplier)
+    }
+
+    ProductSupplier updateDetails(ProductSupplierDetailsCommand command, String productSupplierId) {
+        validateProductSupplierDetails(command)
+        ProductSupplier productSupplier = productSupplierGormService.get(productSupplierId)
+        if (!productSupplier) {
+            throw new ObjectNotFoundException(productSupplierId, ProductSupplier.class.toString())
+        }
+        productSupplier.properties = command.properties
+        // TODO: To be replaced by a generic method created in OBPIH-6017 after merging into develop
```

Comment:
as you can see, the if statement below is repeated here, and in the `saveDetails` method. I've implemented a method for that recently, while working on OBPIH-6017 (assignSourceCode): https://github.com/openboxes/openboxes/pull/4486
that we will be able to use here as soon as the feature branch is merged into develop (upgrade to grails 3.3.10) branch.

Reply:
I'd like to keep the code below if possible i.e. revert the assignSourceCode. I like the idea of abstracting the generation and assignment, but I'm not yet comfortable with how it's done in #4486 so we can postpone that for now.